crates/red_knot_python_semantic/src/types.rs

```diff
@@ -542,14 +607,21 @@ impl<'db> Type<'db> {
             Type::Any | Type::Todo | Type::Never | Type::Unknown | Type::Unbound => {
                 Truthiness::Ambiguous
             }
-            Type::None => Truthiness::AlwaysFalse,
-            Type::Function(_) => Truthiness::AlwaysTrue,
-            Type::Module(_) => Truthiness::AlwaysTrue,
+            Type::None | Type::Falsy => Truthiness::AlwaysFalse,
+            Type::Function(_) | Type::Module(_) | Type::Truthy => Truthiness::AlwaysTrue,
             Type::Class(_) => {
                 // TODO: lookup `__bool__` and `__len__` methods on the class's metaclass
                 // More info in https://docs.python.org/3/library/stdtypes.html#truth-value-testing
                 Truthiness::Ambiguous
             }
+            // Temporary special case for `None` until we handle generic instances
+            Type::Instance(class) if class.is_known(db, KnownClass::NoneType) => {
+                Truthiness::AlwaysFalse
+            }
```

Comment:
We currently always treat `None` values as `Type::None` rather than as `Type::Instance(NoneType)` (although we [probably shouldn't](https://github.com/astral-sh/ruff/issues/13670) -- but that might be tricky to resolve until [a few other things](https://github.com/astral-sh/ruff/issues/12694) get done).

None of your tests fail if I remove this branch, so it seems at least to be uncovered:

```suggestion
```

Reply:
No problem with removing this - I also started looking into #13670, and it's actually more work than we would think (`KnownClass::NoneType.to_instance(db)` is a `Type::Union` including two `None` and an `Unknown` ...)
So anyway - working with `Type::None` for now is probably the better choice.

Reply:
This comment about [sealed types](https://github.com/astral-sh/ruff/issues/12694) make me think there might be a connection to the logic I've been trying to cover in this PR